src/vos/vos_ts.h

```diff
@@ -421,6 +432,48 @@ vos_ts_get_negative(struct vos_ts_set *ts_set, uint64_t hash, bool reset)
 	return neg_entry;
 }
 
+/** Do an uncertainty check on the entry.  Return true if there
+ *  is a write within the epoch uncertainty bound or if it
+ *  can't be determined that the epoch is safe (e.g. a cache miss).
+ *
+ *  \param[in]	ts_set	The timestamp set
+ *  \param[in]	epoch	The epoch of the update
+ *  \param[in]	bound	The uncertainty bound
+ */
+static inline bool
+vos_ts_wcheck(struct vos_ts_set *ts_set, daos_epoch_t epoch,
+	      daos_epoch_t bound)
+{
+	struct vos_wts_cache	*wcache;
+	struct vos_ts_set_entry	*se;
+	uint32_t		 max_idx;
+	daos_epoch_t		 max;
+	daos_epoch_t		 min;
+
+	if (!vos_ts_in_tx(ts_set) || ts_set->ts_init_count == 0 ||
+	    bound <= epoch)
+		return false;
+
+	se = &ts_set->ts_entries[ts_set->ts_init_count - 1];
+
+	if (se->se_entry == NULL)
+		return false;
+
+	wcache = &se->se_entry->te_w_cache;
+	max_idx = wcache->wc_w_max;
+	max = wcache->wc_ts_w[max_idx];
+	if (epoch >= max)
+		return false;
+
+	min = wcache->wc_ts_w[1 - max_idx];
+	if (epoch < min) /* unknown case, not enough history */
+		return true;
+
+	if (bound >= max) /* write conflict */
+		return true;
```

Comment:
Did you mean

    if (bound >= min)

instead?

Reply:
no, I added a comment inline in the code...is it clearer ?

Reply:
Yes, much clearer. Thanks.